journal/src/main/java/io/camunda/zeebe/journal/file/JournalSegment.java

```diff
@@ -173,23 +188,55 @@ public boolean isOpen() {
   /** Closes the segment. */
   @Override
   public void close() {
-    writer.close();
+    open = false;
     readers.forEach(MappedJournalSegmentReader::close);
     IoUtil.unmap(buffer);
-    open = false;
   }
 
   /** Deletes the segment. */
   public void delete() {
+    open = false;
+    markForDeletion();
+    if (readers.isEmpty()) {
+      safeDelete();
+    }
+  }
+
+  private void safeDelete() {
+    if (!readers.isEmpty()) {
+      throw new JournalException(
+          String.format(
+              "Cannot delete segment file. There are %d readers referring to this segment.",
+              readers.size()));
+    }
     try {
-      Files.deleteIfExists(file.file().toPath());
+      IoUtil.unmap(buffer);
+      Files.deleteIfExists(file.getFileMarkedForDeletion());
     } catch (final IOException e) {
-      throw new JournalException(e);
+      LOG.warn(
+          "Could not delete segment {}. File to delete {}. This can lead to increased disk usage.",
+          this,
+          file.getFileMarkedForDeletion(),
+          e);
     }
   }
 
   @Override
   public String toString() {
     return toStringHelper(this).add("id", id()).add("index", index()).toString();
   }
+
+  private void markForDeletion() {
+    if (markedForDeletion) {
+      return;
+    }
```

Comment:
:upside_down_face: I find it helps to visually separate the guard clause from the rest of the code, i.e. with an empty line. So you can quickly skip over them to read the actual "processing" code.